amr-wind/turbulence/LES/Kosovic.cpp

```diff
@@ -0,0 +1,195 @@
+    // a temporary buffer
+    fvm::strainrate(mu_turb, vel);
+    // Non-linear component Nij is computed here and goes into Body Forcing
+    // fvm::gradient(m_gradU,vel);
+    fvm::nonlinearsum(m_Nij, vel);
+    fvm::divergence(m_divNij, m_Nij);
+    const int nlevels = repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+        const auto& geom = geom_vec[lev];
+        const auto& problo = repo.mesh().Geom(lev).ProbLoArray();
+        const amrex::Real dx = geom.CellSize()[0];
+        const amrex::Real dy = geom.CellSize()[1];
+        const amrex::Real dz = geom.CellSize()[2];
+        const amrex::Real ds = std::cbrt(dx * dy * dz);
+        const amrex::Real ds_sqr = ds * ds;
+        const amrex::Real smag_factor = Cs_sqr * ds_sqr;
+
+        for (amrex::MFIter mfi(mu_turb(lev)); mfi.isValid(); ++mfi) {
+            const auto& bx = mfi.tilebox();
+            const auto& mu_arr = mu_turb(lev).array(mfi);
+            const auto& rho_arr = den(lev).const_array(mfi);
+            const auto& divNijLevel = (this->m_divNij)(lev).array(mfi);
+            amrex::ParallelFor(
+                bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    const amrex::Real rho = rho_arr(i, j, k);
+                    const amrex::Real x3 = problo[2] + (k + 0.5) * dz;
+                    // if(i==76 && j==76)
+                    // amrex::Print()<<i<<"   "<<j<<"  "<<k<<"  "<<x3<<"
+                    // "<<dz<<std::endl;
+                    const amrex::Real fmu = std::exp(-x3 / m_switchLoc);
+                    amrex::Real phiM = 1;
+                    if (m_refMOL < 0)
+                        phiM = std::pow(1 - 16 * x3 / m_refMOL, -0.25);
+                    else
+                        phiM = 1 + 5 * x3 / m_refMOL;
+                    const amrex::Real ransL =
+                        std::pow(0.41 * (k + 1) * dz / phiM, 2);
+                    amrex::Real turnOff = std::exp(-x3 / m_LESTurnOff);
+                    // if(x3>=m_LESTurnOff)
+                    //   turnOff=0;
+                    // const amrex::Real
+                    // masonScaleInv=1.0/(std::pow(0.41*(k+1)*dz,2))+1/smag_factor;
+                    // amrex::Real
+                    // hybridScale=amrex::min(smag_factor,std::pow(0.41*(k+1)*dz,2));
+                    // amrex::Real
+                    // lscale=m_surfaceFactor*hybridScale+(1-m_surfaceFactor)*smag_factor;
+                    amrex::Real lscale =
+                        m_surfaceFactor *
+                            (std::pow(1 - fmu, m_surfaceRANSExp) * smag_factor +
+                             std::pow(fmu, m_surfaceRANSExp) * ransL) +
+                        (1 - m_surfaceFactor) * smag_factor;
+                    mu_arr(i, j, k) *= rho * lscale * turnOff;
+                    // lscale=m_surfaceFactor*hybridScale+(1-m_surfaceFactor)*smag_factor*0.25*m_C1;
+                    lscale = m_surfaceFactor *
```

Comment:
can we call this another variable? I am worried that the overloaded use of `lscale` would lead to confustion